src/main/java/oshi/hardware/platform/mac/MacSensors.java

```diff
@@ -0,0 +1,305 @@
+            return result;
+        }
+
+        service = IOKit.INSTANCE.IOServiceGetMatchingService(masterPort.getValue(),
+                IOKit.INSTANCE.IOServiceMatching("AppleSMC"));
+        if (service == 0) {
+            LOG.error("Error: no SMC found\n");
+            return result;
+        }
+
+        result = IOKit.INSTANCE.IOServiceOpen(service, SystemB.INSTANCE.mach_task_self(), 0, this.conn);
+        IOKit.INSTANCE.IOObjectRelease(service);
+        if (result != 0) {
+            LOG.error(String.format("Error: IOServiceOpen() = 0x%08x", result));
+            return result;
+        }
+        return 0;
+    }
+
+    private int smcClose() {
+        return IOKit.INSTANCE.IOServiceClose(this.conn.getValue());
+    }
+
+    /*
+     * Get Temperature from SMC
+     */
+    private double smcGetTemperature(String key, int retries) {
+        SMCVal val = new SMCVal();
+        int result = smcReadKey(key, val, retries);
+        if (result == 0) {
+            int temp = 0;
+            if (val.dataSize > 0 && Arrays.equals(val.dataType, IOKit.DATATYPE_SP78)) {
+                temp = (val.bytes[0] * 256 + val.bytes[1]) >> 2;
+            }
+            return temp / 64.0;
+        }
+        // Read failed
+        return 0d;
+    }
+
+    /*
+     * Get number of fans
+     */
+    private int smcGetFanNumber(int retries) {
+        SMCVal val = new SMCVal();
+        int result = smcReadKey(IOKit.SMC_KEY_FAN_NUM, val, retries);
+        if (result == 0) {
+            return strtoul10(val.bytes, val.dataSize);
+        }
+        // Read failed
+        return 0;
+    }
+
+    /*
+     * Get fan speed
+     */
+    private float smcGetFanSpeed(int fanNum, int retries) {
+        SMCVal val = new SMCVal();
+        String key = String.format(IOKit.SMC_KEY_FAN_SPEED, fanNum);
+        int result = smcReadKey(key, val, retries);
+        if (result == 0) {
+            return strtof(val.bytes, val.dataSize, 2);
+        }
+        // Read failed
+        return 0f;
+    }
+
+    /*
+     * Get cached keyInfo if it exists, or generate new
+     */
+    private int smcGetKeyInfo(SMCKeyData inputStructure, SMCKeyData outputStructure) {
+        if (keyInfoCache.containsKey(inputStructure.key)) {
+            outputStructure.keyInfo = keyInfoCache.get(inputStructure.key);
+        } else {
+            int result = 0;
+            inputStructure.data8 = IOKit.SMC_CMD_READ_KEYINFO;
+
+            result = smcCall(IOKit.KERNEL_INDEX_SMC, inputStructure, outputStructure);
+            if (result != 0) {
+                return result;
+            }
+        }
+        return 0;
+    }
+
+    /*
+     * Read a key from SMC
+     */
+    private int smcReadKey(String key, SMCVal val, int retries) {
+        SMCKeyData inputStructure = new SMCKeyData();
+        SMCKeyData outputStructure = new SMCKeyData();
+
+        inputStructure.key = strtoul16(key, 4);
+        int result;
+        do {
+            Util.sleep(8);
```

Comment:
Ummm.... so I see what you meant. I forgot the `break;` at the bottom of that while loop that I had intended!  *facepalm*

Also my caching wasn't working as I forgot my `.put()` ... 

Here's another look at what's happening with the loop break and caching working:
```
100|fail
 98|  1| 14| 16|{"cpuTemperature":62.625,"fanSpeeds":[2454,2458]}
 18|  2|  2|{"cpuTemperature":62.625,"fanSpeeds":[2468,2466]}
  1| 14|  0|{"cpuTemperature":62.75,"fanSpeeds":[2469,2476]}
  0|  2|  0|{"cpuTemperature":62.75,"fanSpeeds":[2488,2479]}
  8|  0| 10|{"cpuTemperature":62.75,"fanSpeeds":[2492,2493]}
  9|  8|  6|{"cpuTemperature":62.625,"fanSpeeds":[2495,2486]}
 19|  1|  0|{"cpuTemperature":62.75,"fanSpeeds":[2482,2492]}
  7|  9|  1|{"cpuTemperature":62.625,"fanSpeeds":[2495,2496]}
  4|  0|  8|{"cpuTemperature":62.5,"fanSpeeds":[2498,2486]}
  0|  0|  0|{"cpuTemperature":62.5,"fanSpeeds":[2502,2495]}
  0|  1|  7|{"cpuTemperature":62.75,"fanSpeeds":[2500,2498]}
```
First call (during initialization) exceeded 100 tries on temp and gave up.
Second call succeeded after 98 tries on temp, 1 on #fans, and 14, 16 on each fan (successfully caching all of the initial dataInfo variables)
Once cached, the subsequent calls were a lot more reliable.